test/gtest/mhp/distributed_dense_matrix3.cpp

```diff
@@ -0,0 +1,214 @@
+  EXPECT_EQ(a.begin()[15], -1);
+  EXPECT_EQ((a.begin()[{4, 4}]), -1);
+  EXPECT_EQ(a.begin()[99], -1);
+  EXPECT_EQ((a.begin()[{10, 10}]), -1);
+
+  // local-only data access
+  if (dr::mhp::default_comm().rank() == 0) {
+    EXPECT_EQ(a[0], -1);
+    EXPECT_EQ(a[15], -1);
+  } else if (dr::mhp::default_comm().rank() == 1) {
+    EXPECT_EQ((a[{4, 4}]), -1);
+  } else /*(dr::mhp::default_comm().rank() == 2) */ {
+    EXPECT_EQ(a[99], -1);
+    EXPECT_EQ((a[{10, 10}]), -1);
+  }
+}
+
+TEST(MhpTests3, DM_For) {
+  assert(dr::mhp::default_comm().size() == 3);
+  const int rows = 11, cols = 11;
+  DM a(rows, cols, -1);
+
+  for (auto i : a) {
+    i = 5;
+  }
+
+  EXPECT_EQ(a.begin()[0], 5);
+  EXPECT_EQ(a.begin()[15], 5);
+  EXPECT_EQ((a.begin()[{4, 4}]), 5);
+  EXPECT_EQ(a.begin()[99], 5);
+  EXPECT_EQ((a.begin()[{10, 10}]), 5);
+
+  if (dr::mhp::default_comm().rank() == 0) {
+    EXPECT_EQ(a[0], 5);
+    EXPECT_EQ(a[15], 5);
+  } else if (dr::mhp::default_comm().rank() == 1) {
+    EXPECT_EQ((a[{4, 4}]), 5);
+  } else /*(dr::mhp::default_comm().rank() == 2) */ {
+    EXPECT_EQ(a[99], 5);
+    EXPECT_EQ((a[{10, 10}]), 5);
+  }
+}
+
+TEST(MhpTests3, DM_Fill) {
+  assert(dr::mhp::default_comm().size() == 3);
+  const int rows = 11, cols = 11;
+  DM a(rows, cols, -13);
+
+  dr::mhp::fill(a, 5);
+
+  EXPECT_EQ(a.begin()[0], 5);
+  EXPECT_EQ(a.begin()[15], 5);
+  EXPECT_EQ((a.begin()[{4, 4}]), 5);
+  EXPECT_EQ(a.begin()[99], 5);
+  EXPECT_EQ((a.begin()[{10, 10}]), 5);
+
+  if (dr::mhp::default_comm().rank() == 0) {
+    EXPECT_EQ(a[0], 5);
+    EXPECT_EQ(a[15], 5);
+  } else if (dr::mhp::default_comm().rank() == 1) {
+    EXPECT_EQ((a[{4, 4}]), 5);
+  } else /*(dr::mhp::default_comm().rank() == 2) */ {
+    EXPECT_EQ(a[99], 5);
+    EXPECT_EQ((a[{10, 10}]), 5);
+  }
+}
+
+TEST(MhpTests3, DM_Fill_HB) {
+  const int rows = 11, cols = 11;
+  dr::mhp::halo_bounds hb(2, 3, false);
+  DM a(rows, cols, -13, hb);
+
+  dr::mhp::fill(a, 5);
+
+  EXPECT_EQ(a.begin()[0], 5);
+  EXPECT_EQ(a.begin()[15], 5);
+  EXPECT_EQ((a.begin()[{4, 4}]), 5);
+  EXPECT_EQ(a.begin()[99], 5);
+  EXPECT_EQ((a.begin()[{10, 10}]), 5);
+
+  if (dr::mhp::default_comm().rank() == 0) {
+    EXPECT_EQ(a[0], 5);
+    EXPECT_EQ(a[15], 5);
+  } else if (dr::mhp::default_comm().rank() == 1) {
+    EXPECT_EQ((a[{4, 4}]), 5);
+  } else /*(dr::mhp::default_comm().rank() == 2) */ {
+    EXPECT_EQ(a[99], 5);
+    EXPECT_EQ((a[{10, 10}]), 5);
+  }
+}
+
+TEST(MhpTests3, DM_Iota_2_args) {
+  const int rows = 11, cols = 11;
+  DM a(rows, cols, -13);
+
+  dr::mhp::iota(a, 0);
+
+  EXPECT_EQ(a.begin()[0], 0);
+  EXPECT_EQ(a.begin()[15], 15);
+  EXPECT_EQ((a.begin()[{4, 4}]), 48);
+  EXPECT_EQ(a.begin()[99], 99);
+  EXPECT_EQ((a.begin()[{10, 10}]), 120);
+
+  if (dr::mhp::default_comm().rank() == 0) {
+    EXPECT_EQ(a[0], 0);
+    EXPECT_EQ(a[15], 15);
+  } else if (dr::mhp::default_comm().rank() == 1) {
+    EXPECT_EQ((a[{4, 4}]), 48);
+  } else /*(dr::mhp::default_comm().rank() == 2) */ {
```

Comment:
`assert(dr::mhp::default_comm().rank() == 2)`
is as descriptive as your comment, but self verifying in addition

here and in other cases